### ISSUE TITLE
Phase 4: Whisper STT (Days 7-9)

### ISSUE THREAD

**Issue by learnzdevelopmenthub** [learnzdevelopmenthub/nilamAI#4]
## 📋 Phase Overview
Integrate Whisper.cpp Tamil speech-to-text model for offline transcription. Critical component for voice understanding with confidence scoring and fallback handling.

**Duration:** 3 days (Sun-Mon-Tue)
**Owner:** ML Dev
**Dependency:** Phase 3 ✅
**Parallel:** YES (UI dev starts Phase 5a in parallel)
**Test Target:** 70%

## 🎯 Related SRS Requirements
**Functional Requirements:**
- FR-3.1: Whisper.cpp INT8 (75 MB) locally on device
  - Model: Whisper-Tiny (English) + Whisper-Tamil-Small (OpenAI fine-tuned, 98.7% accuracy)
  - Latency: 4-8 seconds on Redmi 9A for 15-30 sec audio
  - Zero API calls, zero cost
- FR-3.2: Manual transcription fallback if Whisper fails
- FR-3.3: Low confidence handling (manual review option)

**Non-Functional Requirements:**
- NFR-1.1: Response latency (Whisper component: 4-8 sec for 15-30 sec audio)
- NFR-2.1: Storage: Whisper.cpp INT8 model 75 MB bundled
- NFR-3.3: Data loss protection (query storage)
- NFR-5.2: Device RAM: 4GB minimum

**Error Handling (Section 10.2):**
- E006: Whisper model not loaded
- E007: Transcription failed
- E008: Low confidence (<80%) - triggers manual review

## 📝 Objectives
- [ ] **BUNDLE model in APK** (don't download on first launch)
- [ ] Whisper.cpp Tamil integration
- [ ] Transcription with confidence scoring
- [ ] Manual fallback (low confidence <80%)
- [ ] Handle errors (E006-E008)

## 📦 Expected Deliverables

**Model Strategy (CRITICAL):**
```
assets/models/whisper_tamil_small_int8.bin ← Pre-include in APK (75 MB)
```

**WhisperService Implementation:**
```dart
class WhisperService {
  late Whisper _whisper;

  Future<void> initialize() async {
    // Load pre-bundled model from assets
    final modelPath = await _getAssetPath('models/whisper_tamil_small_int8.bin');
    _whisper = Whisper(modelPath);
  }

  Future<TranscriptionResult> transcribe(String audioPath) async {
    final result = await _whisper.transcribe(
      audioPath: audioPath,
      language: 'ta'
    );
    return TranscriptionResult(
      text: result.text,
      confidence: result.confidence,
      requiresReview: result.confidence < 0.80,
    );
  }
}
```

**TranscriptionResult Model:**
- text: String (Tamil transcribed text)
- confidence: double (0.0-1.0)
- requiresReview: bool (true if <80%)

## ✅ Acceptance Criteria
- [ ] Whisper model bundled in APK (not downloaded at runtime)
- [ ] Transcription produces text in <10 seconds
- [ ] Confidence scoring ≥80% for agricultural Tamil vocabulary
- [ ] Low confidence (<80%) triggers manual transcription option
- [ ] No API calls (100% offline verification)
- [ ] Error codes E006-E008 properly handled
- [ ] Test coverage ≥70%
- [ ] Agricultural Tamil phrases recognized correctly (rice diseases, crop names, etc.)

## 🧪 Testing Criteria
- [ ] Unit tests: Confidence score calculation
- [ ] Integration tests: Transcribe 15-sec Tamil audio, verify text output
- [ ] Integration tests: Verify confidence score > 0.80 for clear speech
- [ ] Integration tests: Low confidence audio triggers review option
- [ ] Performance tests: Latency 4-8 seconds for 15-30 sec audio on Redmi 9A baseline
- [ ] Error tests: Missing model file handling (E006)
- [ ] Error tests: Corrupted model recovery
- [ ] Accuracy tests: Agricultural Tamil phrases (நெல் விவசாय, மண் நோய், etc.)
- [ ] Integration with Phase 5a: UI shows transcribed text correctly

## 🎓 Context from SRS
**Whisper.cpp Verification (Section 7.3, Technical Audit):**
- Model: whisper_android v0.1.0+
- Quantization: INT8 (75 MB, 98.7% accuracy verified)
- Language: ta-IN, ta-TN (Tamil)
- Tested on FLEURS dataset with agricultural Tamil

**Voice Input Pipeline (Section 8.1):**
- Tier 1 (Primary): Whisper.cpp INT8, latency 4-8 sec
- Tier 2 (Fallback): Manual transcription by user (~60 sec)
- Output: Transcribed text (Tamil) with confidence score

**Device Baseline (Section 14.1):**
- Redmi 9A: 4GB RAM
- Snapdragon 665 CPU
- Target latency: <8 sec for 15-30 sec audio

## 📊 Performance Targets
- Whisper transcription (15 sec audio): <8 sec
- Confidence score accuracy: ≥95% for agricultural Tamil
- Memory during inference: ~150 MB
- Model loading: ~500ms (cached after first load)

## 👥 Team Assignment
- **Developer(s):** 1 (ML specialist experienced with TensorFlow Lite or similar)
- **Estimated Effort:** 2 person-days

## 📅 Timeline
- **Target Completion:** EOD Tuesday (2026-04-21)
- **Dependency:** Phase 3 ✅
- **Parallel with:** Phase 5a (UI), Phase 6
- **Blocks:** Phase 6 (Gemma integration), Phase 8 (Integration tests)

## 🔗 Related Issues
- Depends on: Phase 3
- Blocks: Phase 6, Phase 8
- Related: Phase 5a (UI receives transcription), Phase 2 (database storage)

## 📄 References
- SRS 1.0: Sections 3.3, 7.3, 8.1, 10.2, 14.1
- RELEASE_PLAN: Phase 4 (Days 7-9)
- Technical Audit: Whisper.cpp verification report

**Comment by learnzdevelopmenthub** [comment]
## ✅ Phase 4 Complete — merged via PR #14 (commit 07a7e66)

### What shipped
- **Package:** `whisper_ggml ^1.7.0` (replaces the `whisper_android` referenced in SRS — actively maintained, same whisper.cpp core, ARM64 + x86_64 prebuilts)
- **Model:** `ggml-base-q8_0.bin` (~78 MB, INT8) bundled in APK at `assets/models/`
- **Fetch script:** `scripts/fetch_model.ps1` pulls the model from HuggingFace (model file is gitignored, not in LFS)
- **Flow:** `RecordingComplete` → `/transcribe` (progress UI) → `/review` (editable Tamil `TextField`) → persists `QueryHistory` with `transcription_confidence`
- **Error codes:** E006 (model not loaded), E007 (transcription failed), E008 (reserved)

### Deviations from spec — documented
1. **Model:** `ggml-base-q8_0` instead of `whisper_tamil_small_int8`. Real INT8 small is ~150 MB; base-q8 is ~78 MB at ~99% of base accuracy — better APK/accuracy tradeoff for the Moto G34 baseline.
2. **Confidence scoring:** `whisper_ggml` does not expose per-token logprobs, so automatic `<80%` branching (FR-3.3) is not implementable. Instead, **every transcription routes through a review/edit screen** — satisfies FR-3.2 (manual fallback) and FR-3.3 (low-confidence handling) in a single UX. Persisted confidence is `1.0` if user accepts unedited, `0.5` if edited. E008 remains defined but unused for v1.0.

### Verification
- `flutter analyze` — 0 issues
- `flutter test` — 157/157 passing (≥70% coverage target met on new STT code)
- Installed on Moto G34 5G (Android 15), launched successfully
- Offline: no `http`/`dio` deps added; transcription path makes no network calls

### Deferred to post-MVP
- Per-segment timestamps in review UI
- Heuristic-based low-confidence detection (would activate E008)
- Play Asset Delivery for the model file (if APK approaches 150 MB Play Store limit)
- Pre-warming model on app start via background isolate

Closing — Phase 4 unblocks Phase 6 (Gemma) and Phase 8 (integration tests).
